File: backend/tiggle/src/main/java/com/side/tiggle/domain/transaction/dto/resp/TransactionRespDto.java
package com.side.tiggle.domain.transaction.dto.resp;

import com.side.tiggle.domain.transaction.dto.TransactionDto;
import com.side.tiggle.domain.transaction.model.Transaction;
import lombok.*;
import lombok.experimental.SuperBuilder;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;

import java.time.LocalDateTime;
import java.util.stream.Collectors;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@SuperBuilder
public class TransactionRespDto extends TransactionDto {

    private Long id;
    private TransactionDto parentTx;
    private LocalDateTime createdAt;
    private Long asset;
    private Long category;
    private String txTagNames;


    public static TransactionRespDto fromEntity(Transaction tx) {
        return TransactionRespDto.builder()
                .id(tx.getId())
                .memberId(tx.getMemberId())
                .parentId(tx.getParentId())
                .type(tx.getType())
                .imageUrl(tx.getImageUrl())
                .amount(tx.getAmount())
                .date(tx.getDate())
                .content(tx.getContent())
                .reason(tx.getReason())
                .createdAt(tx.getCreatedAt())
                .build();
    }

    public static TransactionRespDto fromEntityDetailTx(Transaction tx, Long assetId, Long categoryId, String txTagNames) {
        return TransactionRespDto.builder()
                .id(tx.getId())
                .memberId(tx.getMemberId())
                .parentId(tx.getParentId())
                .type(tx.getType())
                .imageUrl(tx.getImageUrl())
                .amount(tx.getAmount())
                .date(tx.getDate())
                .content(tx.getContent())
                .reason(tx.getReason())
                .asset(assetId)
                .category(categoryId)
                .txTagNames(txTagNames)
                .createdAt(tx.getCreatedAt())
                .build();
    }

    public static TransactionRespDto fromEntityParentTx(Transaction refundTx, Transaction parentTx, Long assetId, Long categoryId, String txTagNames) {
        return TransactionRespDto.builder()
                .id(refundTx.getId())
                .memberId(refundTx.getMemberId())
                .parentId(refundTx.getParentId())
                .type(refundTx.getType())
                .imageUrl(refundTx.getImageUrl())
                .amount(refundTx.getAmount())
                .date(refundTx.getDate())
                .content(refundTx.getContent())
                .reason(refundTx.getReason())
                .asset(assetId)
                .category(categoryId)
                .txTagNames(txTagNames)
                .createdAt(refundTx.getCreatedAt())
                .parentTx(TransactionDto.fromEntity(parentTx))
                .build();
    }

    public static Page<TransactionRespDto> fromEntityPage(Page<Transaction> txPage) {
        return new PageImpl<>(
                txPage.getContent()
                        .stream()
                        .map(TransactionRespDto::fromEntity)
                        .collect(Collectors.toList()),
                txPage.getPageable(),
                txPage.getTotalElements()
        );
    }
}

File: backend/tiggle-root/tiggle/src/main/kotlin/com/side/tiggle/domain/transaction/dto/resp/TransactionRespDto.kt
package com.side.tiggle.domain.transaction.dto.resp

import com.side.tiggle.domain.category.dto.resp.CategoryRespDto
import com.side.tiggle.domain.member.dto.resp.MemberRespDto
import com.side.tiggle.domain.transaction.model.Transaction
import java.time.LocalDate
import java.time.LocalDateTime

data class TransactionRespDto(
    val id: Long,
    val member: MemberRespDto,
    val category: CategoryRespDto,
    val type: String = "OUTCOME",
    val tagNames: List<String>?,
    val createdAt: LocalDateTime,
    val parentId: Long? = null,
    val imageUrl: String?,
    val amount: Int,
    val date: LocalDate,
    val content: String,
    val reason: String,
) {
    companion object {
        fun fromEntity(tx: Transaction): TransactionRespDto {
            return TransactionRespDto(
                id = tx.id!!,
                member = MemberRespDto.fromEntity(tx.member),
                category = CategoryRespDto.fromEntity(tx.category),
                imageUrl = tx.imageUrl,
                amount = tx.amount,
                date = tx.date,
                content = tx.content,
                reason = tx.reason,
                createdAt = tx.createdAt!!,
                tagNames = tx.tagNames
            )
        }
    }

}
